airflow/contrib/operators/kubernetes_operator.py

```diff
@@ -172,26 +173,72 @@ def log_container_logs(self, job_name):
-            matched = re.search(r'(\d+) Running / \d+ Succeeded / (\d+) Failed', job_description)
-            logging.info('Current status is: %s' % matched.group(0))
-
-            running_job_count = int(matched.group(1))
-            failed_job_count = int(matched.group(2))
-            if failed_job_count != 0:
-                self.log_container_logs(job_name)
-                raise Exception('%s has failed pods, failing task.' % job_name)
+        pod_output = None  # keeping this out here so we can reuse it in the "finally" clause
+
+        try:
+            while True:
+                time.sleep(self.sleep_seconds_between_polling)
+
+                pod_output = json.loads(subprocess.check_output(args=[
+                    'kubectl', 'get', 'pods', '-o', 'json', '-l', 'job-name==%s' % job_name]))
+                job_description = subprocess.check_output(args=['kubectl', 'describe', 'job', job_name])
+                matched = re.search(r'(\d+) Running / \d+ Succeeded / (\d+) Failed', job_description)
+                logging.info('Current status is: %s' % matched.group(0))
+
+                running_job_count = int(matched.group(1))
+                failed_job_count = int(matched.group(2))
+                if failed_job_count != 0:
+                    self.log_container_logs(job_name)
+                    raise Exception('%s has failed pods, failing task.' % job_name)
+
+                if running_job_count == 0:
+                    return
+
+                # Determine if we have any containers left running in each pod of the job.
+                # Dependent containers don't count.
+                # If there are no pods left running anything, we are done here. Cleaning up
+                # dependent containers will be left to the top-level `finally` block down below.
+                live_pods = 0
+                for pod in pod_output['items']:
+                    # exclude non-running pods
+                    if 'Running' != pod['status']['phase']:
+                        continue
+
+                    live_pods += 1
+                    # get all of the independent containers that are still alive
+                    independent_live = [
+                        cs['name']
+                        for cs
+                        in pod['status']['containerStatuses']
+                        if 'running' in cs['state'] and cs['name'] not in dependent_containers
+                    ]
+                    # if there are none, consider the pod dead
+                    if len(independent_live) == 0:
+                        live_pods -= 1
+
+                # we have no live pods. end the job and return ok
+                if live_pods == 0:
+                    logging.info('No live, independent pods left. Exiting poll loop.')
+                    return
+        finally:
+            if pod_output:
+                # let's clean up all our old pods. we'll kill the entry point (PID 1) in each running container
```

Comment:
If we exit the loop because of failure we don't run our clean_up. That's intentional -- we don't want to _delete_ old jobs that fail, but we don't want to leave a bunch of running `cloudsql-proxy` containers lying around either

Reply:
ah, I think we're not on the same page about old jobs!

I think we do want to delete all old jobs now that we push the logs to GCS (since people will be able to see the reason for failure there). Is there another reason to keep them around that I'm not thinking of?

If there is, we should create a ticket to clean up dead jobs after some time period.